### DISCUSSION TITLE
GFS gribs in S3: Scan grib and fast referencing using index file, streaming it with Dask cluster in Coiled

### DISCUSSION THREAD

**Discussion by nishadhka** [fsspec/kerchunk#530] [Show and tell]
Thanks to kerchunk scan_grib and the method in fast referencing [dynamic-Grib-chunking method](https://github.com/asascience-open/nextgen-dmac/commit/6b3286627070c36127ec97b7dbbb88b0ab481f06), could able to make GFS grib references into paraquet file for a single 00z run. 
As the issue with duplicated references https://github.com/fsspec/kerchunk/issues/407, a list of variables and its levels are passed to make the references, which looks avoiding the issue of referencing and zarr creation. 
The file [run_day_gfs_gik.py](https://github.com/icpac-igad/grib-index-kerchunk/blob/deef5a92ba8cf57aec1b8105af4029b34ccee9d3/run_day_gfs_gik.py) is used to create the parquet file with custom functions from [utils.py](https://github.com/icpac-igad/grib-index-kerchunk/blob/deef5a92ba8cf57aec1b8105af4029b34ccee9d3/utils.py). There is pre made scan_grib file for each time step is stored as paraquet and used in during the process. Which makes the single referencing for next 120 hours about 2 minutes to complete in a Coiled notebook in same region of GFS S3 bucket. 

The file [run_day_stream_gfs_gik_to_zarr.py](https://github.com/icpac-igad/grib-index-kerchunk/blob/deef5a92ba8cf57aec1b8105af4029b34ccee9d3/run_day_stream_gfs_gik_to_zarr.py) reads the created paraquet into zarr and stream 15 variables nearly 15GB in size under 5 minutes. 

The parquet references files for 15 variables are stored in https://huggingface.co/datasets/Nishadhka/gfs_s3_gik_refs/tree/main

**Comment by martindurant** [comment]
cc @emfdavid

**Comment by emfdavid** [comment]
So glad you found this work helpful!
We are moving the dynamic chunking code from the prototype asascience-open repo to kerchunk.
The api is now complete in kerchunk. The tests are still in progress ([PR](https://github.com/fsspec/kerchunk/pull/528))!

**Comment by emfdavid** [comment]
Can you share more about your work and what you are doing with the data?

**Comment by nishadhka** [comment]
This is related to using Ensemble Prediction System(EPS) (GEFS and ECMWF) for impact based forecasting and anticipatory action, especially for disasters such as floods and drought. The data is being used to train ML models to be used for EPS post processing and bias correction. 

@emfdavid Thanks for the heads up on testing. Planning for more tests, on ensemble members, on the long term parquet references, data integrity hence storing the references in huggingface.

**Comment by emfdavid** [comment]
@nishadhka all the code, including tests are now merged into kerchunk.
Let me know if you have any issues moving to the supported version.

**Comment by nishadhka** [comment]
@emfdavid, The paraquet creation step working great , but faced issue in converting the parquet reference file into zarr, seems to be issue with Zarr3 integration of datatree. 

Could bale to make the scan_grib, deflate and inflate steps using the functions from dynamic_zarr_Store integrated with Kerchunk. https://github.com/icpac-igad/grib-index-kerchunk/blob/5702e493e7502a187893aa7d4996962c924f8b06/gfs/utils.py#L23

The conversion of parquet into zarr as per the tests in kerchunk https://github.com/fsspec/kerchunk/blob/main/tests/test__grib_idx.py

```
import pandas as pd
import xarray as xr
import zarr
import fsspec
from kerchunk.utils import refs_as_store  # This is a key function

# Load parquet file
parquet_path = 'gfs_20250303_00.par'
zstore_df = pd.read_parquet(parquet_path)
zstore_df['value'] = zstore_df['value'].apply(lambda x: x.decode('utf-8') if isinstance(x, bytes) else x)
zstore = {row['key']: eval(row['value']) for _, row in zstore_df.iterrows()}

# Use refs_as_store to properly handle the Kerchunk references with Zarr 3
store = refs_as_store(zstore)
dt = xr.open_datatree(store, engine="zarr", consolidated=False)
```

Which ends up in error 
```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[1], line 15
     13 # Use refs_as_store to properly handle the Kerchunk references with Zarr 3
     14 store = refs_as_store(zstore)
---> 15 dt = xr.open_datatree(store, engine="zarr", consolidated=False)
.....
821     return await AsyncGroup.from_store(
    822         store_path,
    823         zarr_format=_zarr_format,
    824         overwrite=overwrite,
    825         attributes=attributes,
    826     )
--> 827 raise FileNotFoundError(f"Unable to find group: {store_path}")

FileNotFoundError: Unable to find group: <FsspecStore(ReferenceFileSystem, [/](https://cluster-dajdl.dask.host/))>
```

Method without kerchunk refs_as_Store ends up in following error, considering the FSMAP issue
```
import pandas as pd
import xarray as xr
import fsspec
import zarr


parquet_path='gfs_20250303_00.par'
zstore_df = pd.read_parquet(parquet_path)
zstore_df['value'] = zstore_df['value'].apply(lambda x: x.decode('utf-8') if isinstance(x, bytes) else x)
zstore = {row['key']: eval(row['value']) for _, row in zstore_df.iterrows()}

fs = fsspec.filesystem("reference", fo=zstore)  # Remove asynchronous=True
store = zarr.storage.FsspecStore(fs, path="")
dt = xr.open_datatree(store,engine="zarr", consolidated=False)
```
error with 
```
[/tmp/ipykernel_8470/1833319567.py:13](https://cluster-dajdl.dask.host/tmp/ipykernel_8470/1833319567.py#line=12): UserWarning: fs (<fsspec.implementations.reference.ReferenceFileSystem object at 0x7f7dcb15cd70>) was not created with `asynchronous=True`, this may lead to surprising behavior
  store = zarr.storage.FsspecStore(fs, path="")
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[1], line 14
     12 fs = fsspec.filesystem("reference", fo=zstore)  # Remove asynchronous=True
     13 store = zarr.storage.FsspecStore(fs, path="")
---> 14 dt = xr.open_datatree(store,engine="zarr", consolidated=False)
File [/opt/conda/envs/gik/lib/python3.13/site-packages/zarr/api/asynchronous.py:827](https://cluster-dajdl.dask.host/opt/conda/envs/gik/lib/python3.13/site-packages/zarr/api/asynchronous.py#line=826), in open_group(store, mode, cache_attrs, synchronizer, path, chunk_store, storage_options, zarr_version, zarr_format, meta_array, attributes, use_consolidated)
    820     _zarr_format = zarr_format or _default_zarr_format()
    821     return await AsyncGroup.from_store(
    822         store_path,
    823         zarr_format=_zarr_format,
    824         overwrite=overwrite,
    825         attributes=attributes,
    826     )
--> 827 raise FileNotFoundError(f"Unable to find group: {store_path}")

FileNotFoundError: Unable to find group: <FsspecStore(ReferenceFileSystem, )>
```

can you confirm this is related to zarr3 version
such as 
1. datatree/xarray zarr3 support https://github.com/pydata/xarray/issues/9984
2. https://github.com/zarr-developers/zarr-python/issues/2706
